File: csharp/tables/text/table_text.cs
//  table_text -- C# port of examples\Vb6\tables\text\table_text.bas
//  Builds a 3x3 table demonstrating cell text alignment (left/center/right x
//  top/center/bottom), draws it, then redraws it with a 90-degree cell
//  orientation. Uses the flat tbl* exports (TPDFTable handle).
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class TableText
{
    static TErrorProc _errCb = ErrProc;

    public static int ErrProc(IntPtr Data, int ErrCode, string ErrMessage, int ErrType)
    {
        Console.WriteLine(ErrMessage);
        return 0;                       // try to continue
    }

    static void Main()
    {
        int timeStart = Environment.TickCount;

        IntPtr pdf = LumasPdf.pdfNewPDF();
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _errCb);

        LumasPdf.pdfCreateNewPDFW(pdf, "");

        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);

        IntPtr tbl = LumasPdf.tblCreateTable(pdf, 3, 3, 500f, 100f);
        LumasPdf.tblSetBoxProperty(tbl, -1, -1, TTableBoxProperty.tbpBorderWidth, 1f, 1f, 1f, 1f);
        LumasPdf.tblSetFontA(tbl, -1, -1, "Arial", LumasPdfConsts.fsRegular, true, TCodepage.cp1252);
        LumasPdf.tblSetFontA(tbl, -1, 1, "Arial", LumasPdfConsts.fsBold, true, TCodepage.cp1252);
        LumasPdf.tblSetGridWidth(tbl, 1f, 1f);

        string txt = "The cell alignment can be set for text, images, and templates...";
        // NOTE: the VB6/Delphi original passes Len = -1 (auto null-terminated length).
        // The LumasPdf engine's tblSetCellTextA crashes (AccessViolation) on Len = -1,
        // so we pass the explicit string length here. (Engine bug: -1 sentinel unhandled.)
        uint all = (uint)txt.Length;

        // -1.0 means use the default row height as specified in the CreateTable() call.
        int rowNum = LumasPdf.tblAddRow(tbl, -1f);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 0, LumasPdfConsts.taLeft, TCellAlign.coTop, txt, all);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 1, LumasPdfConsts.taCenter, TCellAlign.coTop, txt, all);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 2, LumasPdfConsts.taRight, TCellAlign.coTop, txt, all);

        rowNum = LumasPdf.tblAddRow(tbl, -1f);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 0, LumasPdfConsts.taLeft, TCellAlign.coCenter, txt, all);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 1, LumasPdfConsts.taCenter, TCellAlign.coCenter, txt, all);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 2, LumasPdfConsts.taRight, TCellAlign.coCenter, txt, all);

        rowNum = LumasPdf.tblAddRow(tbl, -1f);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 0, LumasPdfConsts.taLeft, TCellAlign.coBottom, txt, all);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 1, LumasPdfConsts.taCenter, TCellAlign.coBottom, txt, all);
        LumasPdf.tblSetCellTextA(tbl, (uint)rowNum, 2, LumasPdfConsts.taRight, TCellAlign.coBottom, txt, all);

        // Draw the table now
        LumasPdf.pdfAppend(pdf);
        LumasPdf.tblDrawTable(tbl, 50f, 50f, 742f);
        while (LumasPdf.tblHaveMore(tbl))
        {
            LumasPdf.pdfEndPage(pdf);
            LumasPdf.pdfAppend(pdf);
            LumasPdf.tblDrawTable(tbl, 50f, 50f, 742f);
        }
        LumasPdf.pdfEndPage(pdf);

        // Let's change the cell orientation to see what happens...
        LumasPdf.tblSetCellOrientation(tbl, -1, -1, 90);
        LumasPdf.pdfAppend(pdf);
        LumasPdf.pdfSetFontW(pdf, "Arial", LumasPdfConsts.fsRegular, 12.0, true, TCodepage.cp1252);
        LumasPdf.pdfWriteTextW(pdf, 50.0, 50.0, "The same table but the cell orientation was changed to 90 degrees.");

        LumasPdf.tblDrawTable(tbl, 50f, 65f, 742f);
        while (LumasPdf.tblHaveMore(tbl))
        {
            LumasPdf.pdfEndPage(pdf);
            LumasPdf.pdfAppend(pdf);
            LumasPdf.tblDrawTable(tbl, 50f, 50f, 737f);
        }
        LumasPdf.pdfEndPage(pdf);

        LumasPdf.tblDeleteTable(ref tbl);   // frees the table and sets tbl to 0

        // A table stores errors and warnings in the error log
        TPDFError err = new TPDFError();
        err.StructSize = (uint)Marshal.SizeOf(typeof(TPDFError));
        for (int i = 0; i < LumasPdf.pdfGetErrLogMessageCount(pdf); i++)
        {
            LumasPdf.pdfGetErrLogMessage(pdf, (uint)i, ref err);
            Console.WriteLine(Marshal.PtrToStringAnsi(err.Msg));
        }

        // No fatal error occurred?
        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            // We write the file into the application directory.
            string outFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile))
            {
                LumasPdf.pdfDeletePDF(pdf);
                return;
            }
            if (LumasPdf.pdfCloseFile(pdf))
            {
                timeStart = Environment.TickCount - timeStart;
                Console.WriteLine("Processing time: " + timeStart + " ms");
            }
        }

        LumasPdf.pdfDeletePDF(pdf);
    }
}

File: cpp/tables/text/table_text.cpp
// table_text -- C++ mirror of examples\Vb6\tables\text\table_text.bas
// Builds a 3x3 table demonstrating cell text alignment, draws it, then redraws
// with a 90-degree cell orientation. Uses the flat tbl* exports.
#include "apputil.h"

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType) {
    if (ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

int main() {
    ChdirToExe();
    unsigned long timeStart = GetTickCount();

    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, 0, ErrProc);
    pdfCreateNewPDFA(pdf, "");
    pdfSetPageCoords(pdf, pcTopDown);

    ITBL tbl = tblCreateTable(pdf, 3, 3, 500.0f, 100.0f);
    tblSetBoxProperty(tbl, -1, -1, tbpBorderWidth, 1, 1, 1, 1);
    tblSetFontA(tbl, -1, -1, "Arial", fsRegular, 1, cp1252);
    tblSetFontA(tbl, -1, 1, "Arial", fsBold, 1, cp1252);
    tblSetGridWidth(tbl, 1, 1);

    const char* txt = "The cell alignment can be set for text, images, and templates...";

    int rowNum = tblAddRow(tbl, -1.0f);
    tblSetCellTextA(tbl, rowNum, 0, taLeft,   coTop, txt, (UI32)strlen(txt));
    tblSetCellTextA(tbl, rowNum, 1, taCenter, coTop, txt, (UI32)strlen(txt));
    tblSetCellTextA(tbl, rowNum, 2, taRight,  coTop, txt, (UI32)strlen(txt));

    rowNum = tblAddRow(tbl, -1.0f);
    tblSetCellTextA(tbl, rowNum, 0, taLeft,   coCenter, txt, (UI32)strlen(txt));
    tblSetCellTextA(tbl, rowNum, 1, taCenter, coCenter, txt, (UI32)strlen(txt));
    tblSetCellTextA(tbl, rowNum, 2, taRight,  coCenter, txt, (UI32)strlen(txt));

    rowNum = tblAddRow(tbl, -1.0f);
    tblSetCellTextA(tbl, rowNum, 0, taLeft,   coBottom, txt, (UI32)strlen(txt));
    tblSetCellTextA(tbl, rowNum, 1, taCenter, coBottom, txt, (UI32)strlen(txt));
    tblSetCellTextA(tbl, rowNum, 2, taRight,  coBottom, txt, (UI32)strlen(txt));

    pdfAppend(pdf);
    tblDrawTable(tbl, 50, 50, 742);
    while (tblHaveMore(tbl) != 0) {
        pdfEndPage(pdf);
        pdfAppend(pdf);
        tblDrawTable(tbl, 50, 50, 742);
    }
    pdfEndPage(pdf);

    // Change the cell orientation to see what happens.
    tblSetCellOrientation(tbl, -1, -1, 90);
    pdfAppend(pdf);
    pdfSetFontA(pdf, "Arial", fsRegular, 12.0, 1, cp1252);
    pdfWriteTextA(pdf, 50, 50, "The same table but the cell orientation was changed to 90 degrees.");
    tblDrawTable(tbl, 50, 65, 742);
    while (tblHaveMore(tbl) != 0) {
        pdfEndPage(pdf);
        pdfAppend(pdf);
        tblDrawTable(tbl, 50, 50, 737);
    }
    pdfEndPage(pdf);

    tblDeleteTable(&tbl);

    TPDFError err; err.StructSize = sizeof(err);
    for (int n = 0; n < (int)pdfGetErrLogMessageCount(pdf); ++n) {
        pdfGetErrLogMessage(pdf, n, &err);
        if (err.Msg) printf("%s\n", err.Msg);
    }

    if (pdfHaveOpenDoc(pdf) != 0) {
        const char* outFile = "out.pdf";
        if (pdfOpenOutputFileA(pdf, outFile) == 0) { pdfDeletePDF(pdf); return 0; }
        if (pdfCloseFile(pdf) != 0) {
            timeStart = GetTickCount() - timeStart;
            printf("Processing time: %lu ms\n", timeStart);
        }
    }

    pdfDeletePDF(pdf);
    return 0;
}
